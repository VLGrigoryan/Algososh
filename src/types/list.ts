import { ElementStates } from "./element-states";

export interface ICircle {
  value: string;
  type: "top" | "bottom";
}

export interface IListArrItem {
  value: string;
  circle: ICircle | undefined;
  state: ElementStates;
}

export interface IStateLoading {
  addToHead: boolean;
  addToTail: boolean;
  deleteFromHead: boolean;
  deleteFromTail: boolean;
  addByIndex: boolean;
  deleteByIndex: boolean;
}

export class Node<T> {
  value: T;
  next: Node<T> | null;
  constructor(value: T, next: Node<T> | null = null) {
    this.value = value;
    this.next = next;
  }
}

export interface INodeList<T> {
  append: (item: T) => void;
  prepend: (item: T) => void;
  deleteFromHead: () => void;
  deleteFromTail: () => void;
  addByIndex: (item: T, index: number) => void;
  deleteByIndex: (index: number) => void;
  getSize: () => number;
  getArr: () => T[];
}