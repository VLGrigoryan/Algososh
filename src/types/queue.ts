export interface IQueue<T> {
    enqueue(item: T): void;
    dequeue(): void;
    clear(): void;
    getHead(): number;
    getTail(): number;
    getSize(): number;
    getQueue(): Array<T | undefined>;
    getLength(): number;
    isEmpty(): boolean;
  }
  