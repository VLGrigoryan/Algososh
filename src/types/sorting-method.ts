export enum SortingMethod {
    Selection = "selection",
    Bubble = "bubble",
  }