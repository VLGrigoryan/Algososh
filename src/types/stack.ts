import { ElementStates } from "./element-states";

export interface IElements {
  value: string;
  state: ElementStates;
  isHead: boolean;
  isTail: boolean;
}

export interface IStack<T> {
  push: (item: T) => void;
  pop: () => T | undefined;
  clear: () => void;
  getElements: () => T[];
  getSize: () => number;
}

export class Stack<T> implements IStack<T> {
  private container: T[] = [];
  push = (item: T): void => {
    this.container.push(item);
  };
  pop = (): T | undefined => this.container.pop();
  clear = (): void => {
    this.container.length = 0;
  };
  getElements = (): T[] => this.container;
  getSize = (): number => this.container.length;

  initializeWithSize(size: number): void {
    this.container = new Array<T>(size);
  }
}
