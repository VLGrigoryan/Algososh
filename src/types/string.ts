import { ElementStates } from "./element-states";

export interface ILetter<T> {
    value: T;
    state: ElementStates;
    isHead?: boolean;
    isTail?: boolean;
  }
  