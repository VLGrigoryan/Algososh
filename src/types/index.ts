import { Stack } from "./stack";
import { Direction } from "./direction";
import { SortingMethod } from "./sorting-method";
import { ElementStates } from "./element-states";
 export  { Stack, SortingMethod, Direction, ElementStates } 
