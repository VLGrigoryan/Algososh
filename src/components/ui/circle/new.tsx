import { cleanup, render, screen, within } from "@testing-library/react";
import { ElementStates } from "../../../types/element-states";
import { Circle } from "./circle";

describe("Circle component renders correctly", () => {
  afterEach(cleanup);

  it("without letter", () => {
    render(<Circle />);
    const circle = screen.getByTestId("circle");
    const letter = within(circle).getByTestId("letter");
    expect(letter).toHaveTextContent("");
    expect(circle).toMatchSnapshot();
  });

  it("with letter", () => {
    render(<Circle letter="a32d" />);
    const circle = screen.getByTestId("circle");
    const letter = within(circle).getByTestId("letter");
    expect(letter).toHaveTextContent("a32d");
    expect(circle).toMatchSnapshot();
  });

  it("with head", () => {
    render(<Circle head="head" />);
    const circle = screen.getByTestId("circle");
    const head = within(circle).getByTestId("head");
    expect(head).toHaveTextContent("head");
    expect(circle).toMatchSnapshot();
  });

  it("with head as react-element", () => {
    render(<Circle head={<Circle isSmall />} />);
    const circle = screen.getAllByTestId("circle")[0];
    const head = within(circle).getAllByTestId("head")[0];
    expect(head).toHaveClass("element");
    expect(circle).toMatchSnapshot();
  });

  it("with tail", () => {
    render(<Circle tail="tail" />);
    const circle = screen.getByTestId("circle");
    const tail = within(circle).getByTestId("tail");
    expect(tail).toHaveTextContent("tail");
    expect(circle).toMatchSnapshot();
  });

  it("with tail as react-element", () => {
    render(<Circle tail={<Circle isSmall />} />);
    const circle = screen.getAllByTestId("circle")[0];
    const tail = within(circle).getAllByTestId("tail")[0];
    expect(tail).toHaveClass("element");
    expect(circle).toMatchSnapshot();
  });

  it("with index", () => {
    render(<Circle index={1} />);
    const circle = screen.getByTestId("circle");
    const index = within(circle).getByTestId("index");
    expect(index).toHaveTextContent("1");
    expect(circle).toMatchSnapshot();
  });

  it("with isSmall", () => {
    render(<Circle isSmall />);
    const circle = screen.getByTestId("circle");
    const core = within(circle).getByTestId("core");
    expect(core).toHaveClass("small");
    expect(circle).toMatchSnapshot();
  });

  it("with default state", () => {
    render(<Circle />);
    const circle = screen.getByTestId("circle");
    const core = within(circle).getByTestId("core");
    expect(core).toHaveClass("default");
    expect(circle).toMatchSnapshot();
  });

  it("with changing state", () => {
    render(<Circle state={ElementStates.Changing} />);
    const circle = screen.getByTestId("circle");
    const core = within(circle).getByTestId("core");
    expect(core).toHaveClass("changing");
    expect(circle).toMatchSnapshot();
  });

  it("with modified state", () => {
    render(<Circle state={ElementStates.Modified} />);
    const circle = screen.getByTestId("circle");
    const core = within(circle).getByTestId("core");
    expect(core).toHaveClass("modified");
    expect(circle).toMatchSnapshot();
  });
});
