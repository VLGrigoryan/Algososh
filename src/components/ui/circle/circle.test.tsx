import React from "react";
import { render, screen, within } from "@testing-library/react";
import { Circle } from "./circle";
import { ElementStates } from "../../../types/element-states";

describe("Circle Component", () => {
  it("should render without letter", () => {
    render(<Circle />);
    expect(screen.getByTestId("circle")).toBeInTheDocument();
    expect(screen.getByTestId("circle")).toMatchSnapshot("Circle without letter");
  });

  it("should render with letter", () => {
    render(<Circle letter="A" />);
    expect(screen.getByText("A")).toBeInTheDocument();
    expect(screen.getByTestId("circle")).toMatchSnapshot("Circle with letter A");
  });

  it("should render with head", () => {
    render(<Circle head="Head" />);
    expect(screen.getByText("Head")).toBeInTheDocument();
    expect(screen.getByTestId("circle")).toMatchSnapshot("Circle with head 'Head'");
  });

  it("should render with react element in head", () => {
    const reactElement = <span>React Element</span>;
    render(<Circle head={reactElement} />);
    expect(screen.getByText("React Element")).toBeInTheDocument();
    expect(screen.getByTestId("circle")).toMatchSnapshot("Circle with React Element in head");
  });

  it("should render with tail", () => {
    render(<Circle tail="Tail" />);
    expect(screen.getByText("Tail")).toBeInTheDocument();
    expect(screen.getByTestId("circle")).toMatchSnapshot("Circle with tail 'Tail'");
  });

  it("should render with react element in tail", () => {
    const reactElement = <span>React Element</span>;
    render(<Circle tail={reactElement} />);
    expect(screen.getByText("React Element")).toBeInTheDocument();
    expect(screen.getByTestId("circle")).toMatchSnapshot("Circle with React Element in tail");
  });

  it("should render with index", () => {
    render(<Circle index={42} />);
    const circle = screen.getByTestId("circle");
    const index = within(circle).getByTestId("index");
    expect(index).toHaveTextContent("42");
    expect(screen.getByTestId("circle")).toMatchSnapshot("Circle with index 42");
  });

  it("should render with isSmall prop as true", () => {
    render(<Circle isSmall={true} />);
    const circle = screen.getByTestId("circle");
    const smallCircle = within(circle).getByTestId("state");

    expect(screen.getByTestId("state")).toHaveClass("small");
    expect(circle).toMatchSnapshot("Circle with isSmall prop as true");
  });
  
  it("should render in 'default' state", () => {
    render(<Circle state={ElementStates.Default} />);
    const circle = screen.getByTestId("circle");
    expect(screen.getByTestId("state")).toHaveClass("default");
    expect(circle).toMatchSnapshot("Circle in default state");
  });

  it("should render in 'changing' state", () => {
    render(<Circle state={ElementStates.Changing} />);
    expect(screen.getByTestId("state")).toHaveClass("changing");
    expect(screen.getByTestId("circle")).toMatchSnapshot("Circle in changing state");
  });

  it("should render in 'modified' state", () => {
    render(<Circle state={ElementStates.Modified} />);
    expect(screen.getByTestId("state")).toHaveClass("modified");
    expect(screen.getByTestId("circle")).toMatchSnapshot("Circle in modified state");
  });
});
