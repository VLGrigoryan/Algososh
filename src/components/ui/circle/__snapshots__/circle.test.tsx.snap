// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Circle Component should render in 'changing' state: Circle in changing state 1`] = `
<div
  class="content "
  data-testid="circle"
>
  <div
    class="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    class="circle   changing"
    data-testid="state"
  >
    <p
      class="text text_type_circle text_color_input letter"
      data-testid="letter"
    />
  </div>
  <p
    class="text text_type_input text_color_input mt-4 absolute index"
    data-testid="index"
  />
  <div
    class="text text_type_input text_color_input mt-4 absolute tail30 element"
    data-testid="tail"
  />
</div>
`;

exports[`Circle Component should render in 'default' state: Circle in default state 1`] = `
<div
  class="content "
  data-testid="circle"
>
  <div
    class="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    class="circle   default"
    data-testid="state"
  >
    <p
      class="text text_type_circle text_color_input letter"
      data-testid="letter"
    />
  </div>
  <p
    class="text text_type_input text_color_input mt-4 absolute index"
    data-testid="index"
  />
  <div
    class="text text_type_input text_color_input mt-4 absolute tail30 element"
    data-testid="tail"
  />
</div>
`;

exports[`Circle Component should render in 'modified' state: Circle in modified state 1`] = `
<div
  class="content "
  data-testid="circle"
>
  <div
    class="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    class="circle   modified"
    data-testid="state"
  >
    <p
      class="text text_type_circle text_color_input letter"
      data-testid="letter"
    />
  </div>
  <p
    class="text text_type_input text_color_input mt-4 absolute index"
    data-testid="index"
  />
  <div
    class="text text_type_input text_color_input mt-4 absolute tail30 element"
    data-testid="tail"
  />
</div>
`;

exports[`Circle Component should render with head: Circle with head 'Head' 1`] = `
<div
  class="content "
  data-testid="circle"
>
  <div
    class="text text_type_input text_color_input mb-4 absolute head string"
  >
    Head
  </div>
  <div
    class="circle   default"
    data-testid="state"
  >
    <p
      class="text text_type_circle text_color_input letter"
      data-testid="letter"
    />
  </div>
  <p
    class="text text_type_input text_color_input mt-4 absolute index"
    data-testid="index"
  />
  <div
    class="text text_type_input text_color_input mt-4 absolute tail30 element"
    data-testid="tail"
  />
</div>
`;

exports[`Circle Component should render with index: Circle with index 42 1`] = `
<div
  class="content "
  data-testid="circle"
>
  <div
    class="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    class="circle   default"
    data-testid="state"
  >
    <p
      class="text text_type_circle text_color_input letter"
      data-testid="letter"
    />
  </div>
  <p
    class="text text_type_input text_color_input mt-4 absolute index"
    data-testid="index"
  >
    42
  </p>
  <div
    class="text text_type_input text_color_input mt-4 absolute tail60 element"
    data-testid="tail"
  />
</div>
`;

exports[`Circle Component should render with isSmall prop as true: Circle with isSmall prop as true 1`] = `
<div
  class="content "
  data-testid="circle"
>
  <div
    class="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    class="circle  small default"
    data-testid="state"
  >
    <p
      class="text text_type_circle text_color_input letter"
      data-testid="letter"
    />
  </div>
  <p
    class="text text_type_input text_color_input mt-4 absolute index"
    data-testid="index"
  />
  <div
    class="text text_type_input text_color_input mt-4 absolute tail30 element"
    data-testid="tail"
  />
</div>
`;

exports[`Circle Component should render with letter: Circle with letter A 1`] = `
<div
  class="content "
  data-testid="circle"
>
  <div
    class="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    class="circle   default"
    data-testid="state"
  >
    <p
      class="text text_type_circle text_color_input letter"
      data-testid="letter"
    >
      A
    </p>
  </div>
  <p
    class="text text_type_input text_color_input mt-4 absolute index"
    data-testid="index"
  />
  <div
    class="text text_type_input text_color_input mt-4 absolute tail30 element"
    data-testid="tail"
  />
</div>
`;

exports[`Circle Component should render with react element in head: Circle with React Element in head 1`] = `
<div
  class="content "
  data-testid="circle"
>
  <div
    class="text text_type_input text_color_input mb-4 absolute head element"
  >
    <span>
      React Element
    </span>
  </div>
  <div
    class="circle   default"
    data-testid="state"
  >
    <p
      class="text text_type_circle text_color_input letter"
      data-testid="letter"
    />
  </div>
  <p
    class="text text_type_input text_color_input mt-4 absolute index"
    data-testid="index"
  />
  <div
    class="text text_type_input text_color_input mt-4 absolute tail30 element"
    data-testid="tail"
  />
</div>
`;

exports[`Circle Component should render with react element in tail: Circle with React Element in tail 1`] = `
<div
  class="content "
  data-testid="circle"
>
  <div
    class="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    class="circle   default"
    data-testid="state"
  >
    <p
      class="text text_type_circle text_color_input letter"
      data-testid="letter"
    />
  </div>
  <p
    class="text text_type_input text_color_input mt-4 absolute index"
    data-testid="index"
  />
  <div
    class="text text_type_input text_color_input mt-4 absolute tail30 element"
    data-testid="tail"
  >
    <span>
      React Element
    </span>
  </div>
</div>
`;

exports[`Circle Component should render with tail: Circle with tail 'Tail' 1`] = `
<div
  class="content "
  data-testid="circle"
>
  <div
    class="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    class="circle   default"
    data-testid="state"
  >
    <p
      class="text text_type_circle text_color_input letter"
      data-testid="letter"
    />
  </div>
  <p
    class="text text_type_input text_color_input mt-4 absolute index"
    data-testid="index"
  />
  <div
    class="text text_type_input text_color_input mt-4 absolute tail30 string"
    data-testid="tail"
  >
    Tail
  </div>
</div>
`;

exports[`Circle Component should render without letter: Circle without letter 1`] = `
<div
  class="content "
  data-testid="circle"
>
  <div
    class="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    class="circle   default"
    data-testid="state"
  >
    <p
      class="text text_type_circle text_color_input letter"
      data-testid="letter"
    />
  </div>
  <p
    class="text text_type_input text_color_input mt-4 absolute index"
    data-testid="index"
  />
  <div
    class="text text_type_input text_color_input mt-4 absolute tail30 element"
    data-testid="tail"
  />
</div>
`;
